File: samplejava/src/main/java/com/prof/youtubeparser/sample/java/MainViewModel.java
package com.prof.youtubeparser.sample.java;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.prof.youtubeparser.Parser;
import com.prof.youtubeparser.VideoStats;
import com.prof.youtubeparser.models.stats.Statistics;
import com.prof.youtubeparser.models.videos.Video;

import java.util.ArrayList;
import java.util.List;

public class MainViewModel extends ViewModel {

    private MutableLiveData<List<Video>> mVideoLiveList = new MutableLiveData<>();
    private MutableLiveData<String> mSnackbar = new MutableLiveData<>();
    private MutableLiveData<Statistics> mLiveStats = new MutableLiveData<>();

    private String mNextToken = null;
    private Parser mParser = new Parser();
    private ArrayList<Video> mVideoList = new ArrayList<>();
    private final String CHANNEL_ID = "UCVHFbqXqoYvEWM1Ddxl0QDg";

    public MutableLiveData<List<Video>> getVideoLiveList() {
        return mVideoLiveList;
    }

    private void setVideoLiveList(List<Video> videoList) {
        this.mVideoLiveList.postValue(videoList);
    }

    public MutableLiveData<String> getSnackbar() {
        return mSnackbar;
    }

    public void onSnackbarShowed() {
        this.mSnackbar.setValue(null);
    }

    public MutableLiveData<Statistics> getLiveStats() {
        return mLiveStats;
    }

    private void setLiveStats(Statistics stats) {
        this.mLiveStats.postValue(stats);
    }

    private void setNextToken(String nextToken) {
        this.mNextToken = nextToken;
    }

    public void fetchVideos() {

        mParser.onFinish(new Parser.OnTaskCompleted() {
            @Override
            public void onTaskCompleted(@NonNull List<Video> list, @NonNull String nextPageToken) {
                setNextToken(nextPageToken);
                mVideoList.addAll(list);
                setVideoLiveList(mVideoList);
            }

            @Override
            public void onError(@NonNull Exception e) {
                setVideoLiveList(new ArrayList<Video>());
                e.printStackTrace();
                mSnackbar.postValue("An error has occurred. Please try again");
            }
        });

        String requestUrl = mParser.generateRequest(
                CHANNEL_ID,
                20,
                Parser.ORDER_DATE,
                BuildConfig.KEY
        );

        mParser.execute(requestUrl);

    }

    public void fetchNextVideos() {
        if (mNextToken != null) {
            mParser.onFinish(new Parser.OnTaskCompleted() {
                @Override
                public void onTaskCompleted(@NonNull List<Video> list, @NonNull String nextPageToken) {
                    setNextToken(nextPageToken);
                    mVideoList.addAll(list);
                    setVideoLiveList(mVideoList);
                }

                @Override
                public void onError(@NonNull Exception e) {
                    e.printStackTrace();
                    mSnackbar.postValue("An error has occurred. Please try again");
                }
            });

            String requestUrl = mParser.generateMoreDataRequest(
                    CHANNEL_ID,
                    20,
                    Parser.ORDER_DATE,
                    BuildConfig.KEY,
                    mNextToken
            );

            mParser.execute(requestUrl);
        }
    }

    public void fetchStatistics(String videoId) {

        VideoStats videoStats = new VideoStats();
        videoStats.onFinish(new VideoStats.OnTaskCompleted() {
            @Override
            public void onTaskCompleted(@NonNull Statistics stats) {
                setLiveStats(stats);
            }

            @Override
            public void onError(@NonNull Exception e) {
                e.printStackTrace();
                mSnackbar.postValue("An error has occurred. Please try again");
            }
        });

        String requestUrl = videoStats.generateStatsRequest(videoId, BuildConfig.KEY);
        videoStats.execute(requestUrl);
    }
}

File: samplekotlin/src/main/java/com/prof/youtubeparser/sample/kotlin/MainViewModel.kt
/*
 *   Copyright 2019 Marco Gomiero
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *
 */

package com.prof.youtubeparser.sample.kotlin

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.prof.youtubeparser.Parser
import com.prof.youtubeparser.VideoStats
import com.prof.youtubeparser.models.stats.Statistics
import com.prof.youtubeparser.models.videos.Video
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

class MainViewModel : ViewModel() {

    private val _videoList = MutableLiveData<List<Video>>()
    val videoList: LiveData<List<Video>>
        get() = _videoList

    private val _snackbar = MutableLiveData<String>()
    val snackbar: LiveData<String>
        get() = _snackbar

    private val _stats = MutableLiveData<Statistics>()
    val stats: LiveData<Statistics>
        get() = _stats

    private var nextToken: String? = null
    private val channelID = "UCVHFbqXqoYvEWM1Ddxl0QDg"
    private val parser = Parser()

    fun onSnackbarShowed() {
        _snackbar.value = null
    }

    fun fetchVideos() {
        viewModelScope.launch(Dispatchers.Main) {
            val requestUrl = parser.generateRequest(
                    channelID = channelID,
                    maxResult = 20,
                    orderType = Parser.ORDER_DATE,
                    key = BuildConfig.KEY
            )
            try {
                val result = parser.getVideos(requestUrl)
                nextToken = result.nextToken
                _videoList.postValue(result.videos)
            } catch (e: Exception) {
                e.printStackTrace()
                _snackbar.value = "An error has occurred. Please retry"
                _videoList.postValue(mutableListOf())
            }
        }
    }

    fun fetchMoreVideos() {
        nextToken?.let { pageToken ->
            val requestUrl = parser.generateMoreDataRequest(
                    channelID = channelID,
                    maxResult = 20,
                    orderType = Parser.ORDER_DATE,
                    key = BuildConfig.KEY,
                    nextToken = pageToken
            )

            viewModelScope.launch {
                val result = parser.getVideos(requestUrl)
                nextToken = result.nextToken
                val oldList = _videoList.value?.toMutableList() ?: mutableListOf()
                oldList.addAll(result.videos)
                _videoList.postValue(oldList)
            }
        }
    }

    fun fetchStatistics(videoId: String) {
        val videoStats = VideoStats()
        val requestUrl = videoStats.generateStatsRequest(
                videoID = videoId,
                key = BuildConfig.KEY
        )
        viewModelScope.launch {
            try {
                _stats.postValue(videoStats.getStats(requestUrl))
            } catch (e: Exception) {
                e.printStackTrace()
                _snackbar.value = "An error has occurred. Please retry"
            }
        }
    }
}